### DISCUSSION TITLE
Stopping Servo Motors via ROS2 Service Call

### DISCUSSION THREAD

**Discussion by lkaising** [Yaskawa-Global/motoros2#319] [Q&A] [answered]
I'm working with the GP4 arm and was wondering if there's a ROS2 service call to stop the servo motors, similar to the `start_traj_mode` service. I attempted to use `stop_traj_mode`, but it doesn't seem to work as expected. Is there a dedicated service to stop the motors, or should I be handling this differently?

Below are some relevant code snippets illustrating what I've tried so far:
```cpp
// Command to start the robot
socket_command_function_map_["startRobot"] = [self = shared_from_this()](std::function<void(bool)> result_callback) {
    RCLCPP_INFO(self->ros_node_instance_->get_logger(), "Start robot command received. Starting robot.");
    auto start_request = std::make_shared<motoros2_interfaces::srv::StartTrajMode::Request>();
    self->start_trajectory_mode_client_->async_send_request(start_request,
    [self, result_callback](rclcpp::Client<motoros2_interfaces::srv::StartTrajMode>::SharedFuture start_future) {
        auto start_result = start_future.get();
        //result_callback(start_result->result_code.value == motoros2_interfaces::msg::MotionReadyEnum::MOTION_READY); 
        // Also, this callback wasn’t working. What should the result code value be?
        result_callback(true);
    });
};
```
The above code for starting the robot works as expected. However, stopping the robot is causing issues. Here’s what I’ve tried:
```cpp
// Command to stop the robot (not working)
socket_command_function_map_["stopRobot"] = [self = shared_from_this()](std::function<void(bool)> result_callback) {
    RCLCPP_INFO(self->ros_node_instance_->get_logger(), "Stop robot command received. Stopping robot.");
    auto stop_request = std::make_shared<std_srvs::srv::Trigger::Request>();
    self->stop_trajectory_mode_client_->async_send_request(stop_request,
    [self, result_callback](rclcpp::Client<std_srvs::srv::Trigger>::SharedFuture stop_future) {
        auto stop_result = stop_future.get();
        result_callback(stop_result->success);
    });
};
````
Unfortunately, the above `stopRobot` command does not seem to stop the robot as expected.

Questions:
1. Is there a dedicated ROS2 service call to stop the motors?
2. Should I be handling the stopping of servo motors differently?
3. Regarding the `start_traj_mode` service, what should the expected `result_code` value be for successful motion readiness?

Any advice or recommendations would be greatly appreciated!

**Comment by gavanderhoorn** [comment]
Are you specifically looking to turn off the servos (ie: power them down), or just want to prevent MotoROS2 from accepting new trajectory goals?

The former is not supported right now (although the servos will disengage after the ECO-mode timeout has passed, depending on whether that's enabled).

The latter is supported, as that's what `stop_traj_mode` does: it causes the `INIT_ROS` INFORM job to terminate which will cause MotoROS2 to not have control over the robot any longer.

As to your questions:

> Is there a dedicated ROS2 service call to stop the motors?

'stop' here is a bit ambiguous, but see my comment above.

> Should I be handling the stopping of servo motors differently?

I'm not sure what you mean by this exactly. Perhaps you could describe what your trying to achieve, just so we can avoid an xy-problem.

> Regarding the `start_traj_mode` service, what should the expected `result_code` value be for successful motion readiness?

That would be `MOTION_READY` in the general case.

If you receive something else, please provide us with a [full debug log](https://github.com/Yaskawa-Global/motoros2/blob/53817ab1a83e81591a02ef23ff990290427052bd/doc/troubleshooting.md#debug-log-client). Make sure to start the debug logger *before* you power on the controller.

You can find the debug logger script in the [tools](https://github.com/Yaskawa-Global/motoros2/tree/53817ab1a83e81591a02ef23ff990290427052bd/tools) sub directory of this repository.

---

Edit:

> but it doesn't seem to work as expected.

please try to explain what it is you observe or what happens and how that is different from what you expect. "doesn't [..] work" is not enough for us to be able to help you.

**Reply by lkaising** [reply]
Thank you for your detailed response. Let me clarify what I meant by "stopping." I am looking to turn off the servo motors, similar to what happens when you press the emergency e-stop button or when you are in play mode on the smart pendant and release your thumb from the black side button. Essentially, I want the high-pitched humming of the robot to stop, indicating that the servos have been powered down.

From what I understand from your explanation, it seems that this feature is not currently supported through a ROS2 service call. This is not ideal for my use case, as manually pressing the e-stop button on the pendant is not practical for more automated control. Down the line, I was hoping to build a solution that allows for this through a ROS2 service. Do you think implementing such a feature is possible, perhaps adding a service similar to `start_traj_mode` that powers down the servo motors?

Regarding the `MOTION_READY` result code, I appreciate the clarification, but the issue I encountered goes beyond recognizing the expected result. When I uncomment the following line:
```cpp
result_callback(start_result->result_code.value == motoros2_interfaces::msg::MotionReadyEnum::MOTION_READY);
```
I receive a "no member" error from the IDE:
```
class "motoros2_interfaces::msg::MotionReadyEnum_<std::allocator<void>>" has no member "MOTION_READY"
```
And when I attempt to build the project, I get this error:
```
error: ‘MOTION_READY’ is not a member of ‘motoros2_interfaces::msg::MotionReadyEnum’ {aka ‘motoros2_interfaces::msg::MotionReadyEnum_<std::allocator<void> >’}
   26 |                 result_callback(start_result->result_code.value == motoros2_interfaces::msg::MotionReadyEnum::MOTION_READY);
```
It seems there may be an issue with how the `MOTION_READY` constant is defined or referenced in the code. Were you aware of this problem, and do you know the correct member to use in this callback function?

Thanks again for your support and insights.

**Reply by gavanderhoorn** [reply, marked as answer]
> From what I understand from your explanation, it seems that this feature is not currently supported through a ROS2 service call.

no, that's indeed not supported right now.

> Do you think implementing such a feature is possible, perhaps adding a service similar to `start_traj_mode` that powers down the servo motors?

no, that would not be too difficult in terms of complexity. We would have to find the time though.

If you have a MotoPlus SDK, you could perhaps consider implementing it and submitting a PR so we can merge it here.

> Regarding the `MOTION_READY` result code, I appreciate the clarification, but the issue I encountered goes beyond recognizing the expected result. When I uncomment the following line:
> 
> ```c++
> result_callback(start_result->result_code.value == motoros2_interfaces::msg::MotionReadyEnum::MOTION_READY);
> ```

ah, ok. I had assumed the issue was with actually receiving the value. My apologies.

The constant is actually called `READY`. See the message definition here: [motoros2_interfaces/msg/MotionReadyEnum.msg](https://github.com/Yaskawa-Global/motoros2_interfaces/blob/855852e62c31ec2a0ed5b2d3735b49e016fbd0ea/msg/MotionReadyEnum.msg#L16-L17).

Does your IDE not auto-complete?

**Reply by lkaising** [reply]
Yup, changing the constant to `READY` resolved the issue. Regarding powering down the servo motors, you confirmed my suspicions, and I really appreciate that. It's not a critical issue for our project right now, but it's definitely something I'd like to explore further. I'll familiarize myself with the MotoPlus SDK and work on this in the background. Thanks again for all your support and guidance!
